### DISCUSSION TITLE
How to create custom sources based on the upstream

### DISCUSSION THREAD

**Discussion by HerbertHe** [HerbertHe/iptv-sources#68] [Q&A] [answered]
How to create custom sources based on the upstream

**Comment by HerbertHe** [comment, marked as answer]
The current version of `iptv-sources` supports create custom sources based on the upstream by yourself.

No matter if you deploy with docker or nodejs, it is very easy to meet the goal.

You can follow the steps:

## Deploy with docker

First, you need to run the container with volume

```shell
docker run -v <host-config-dir-path>:/app/config iptv-sources
```

Then, you just need to add a new `custom.json` file in `<host-config-dir-path>`.

Save, and run `docker exec -d iptv-sources /bin/sh ./update-sources.sh`.

## Deploy with nodejs

Make a new directory `config` in the project root path, add a new `custom.json` file.

Save, and run `yarn m3u`.

**Comment by HerbertHe** [comment]
# How to write the `custom.json` file

The template of `custom.json` file as followed:

```json
{
    "rules": [
        {
            "upstream": "",
            "exclude": [],
            "include": [],
            "append": [
                {
                    "name": "<channel name>",
                    "url": "<url>",
                    "extinf": "<channel tvg information>"
                }
            ],
            "replacer": {
              "extinf": <same as url>,
              "url": [
                    {
                        "pattern": "http://192.168.123.1:23234",
                        "type": "string",
                        "target": "http://192.168.1.1:88"
                    }
                ]
            }
        }
    ]
}
```

| Name | Type | Required | Description |
| --- | --- | --- | --- |
| `rules` | Array | true | Rules for custom sources |
| `upstream` | string | true | The based upstream, like `fmml_ipv6`, `cn`, `j_home` |
| `exclude` | Array\<string\> | false | To exclude the channels |
| `include` | Array\<string\> | false | To include the channels you want, **if the option was configured, the option `exclude` would not work** |
| `append` | Array | false | Append the channels |
| `append.name` | string | true | channel name |
| `append.url` | string | true | channel url |
| `append.extinf` | string | false | channel tvg information |
| `replacer.exinf` `replacer.url` | Array | false | custom replacer |
| `replacer.exinf.type` `replacer.url.type` | `string` or `RegExp` | true | type of replacer |
| `replacer.exinf.pattern` `replacer.url.pattern` | string | true | same as `String.prototype.replace.$0` or `RegExp.pattern` |
| `replacer.exinf.flags` `replacer.url.flags` | string | false | same as [RegExp.flags](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp) |
| `replacer.exinf.target` `replacer.url.target` | string | true | same as [String.prototype.replace.$1](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/replace) |

**Comment by 897896639** [comment]


**Reply by HerbertHe** [reply]
@897896639 去新开issue，附上完整的文件和描述

**Reply by newrookie001** [reply]
能否完整的写一个“custom.json”，这样，我们这些小白用户就能简单的删减或者替换；现在语法格式不太懂，只能按照上面的格式填写，不知道每一个具体数值后面直接跨行还是需要逗号隔开。
希望大神支持一下。最后的重命名在哪里。

**Reply by newrookie001** [reply]
日志显示custom.json的任务执行完成了，但是找不到输出路径，请问该设置哪个参数就出来了。
![Uploading 日志.png…]()
![Uploading custom.png…]()
